Finalize media and rate limit support

weierophinney opened the pull request:
This patch incorporates #34 in order to provide testing. A number of minor refactors were performed to simplify usage, provide consistency, and provide testability.

Additionally, per feedback from @Ocramius, this patch updates the component to target PHP 7.1+, allowing the code to provide scalar type hints and return type hints, which allows far simpler error handling and reduced maintenance overhead.

Ocramius commented:
https://github.com/zendframework/ZendService_Twitter/pull/39/commits/5320b6c1e8ba950556fd76144a694dd102049f1a seems unrelated

A lot of these commits should be squashed due to their lack of description.

Ocramius commented on library/ZendService/Twitter/Image.php:
```diff
@@ -0,0 +1,24 @@
+ */
+class Image extends Media
+{
+    public function __construct($image_url = null, $media_type = 'image/jpeg')
```
Parameter naming is inconsistent with the coding standard

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+    /**
+     * Array of basic data to be stored by this class.
+     *
+     * @var Array
```
`Array` is incorrect - a more specific description is needed

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+
+    /**
+     * The number of bytes to send to Twitter at one time.
+     * @var Integer
```
`int`

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * @param  null|string $imageFilename
+     * @param  string $consumer
+     */
+    public function __construct($imageFilename = null, $media_type = '')
```
For new features, we should *REALLY* use PHP 7.1 as a base for development. That means proper parameter hints. Alternatively, assertions about types need to be added to the code.

Ocramius replied:
Parameter naming does not match CS

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * Constructor
+     *
+     * @param  null|string $imageFilename
+     * @param  string $consumer
```
Docblock inconsistent with constructor arguments

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * Main call into the upload workflow
+     *
+     * @param  Client $httpClient
+     * @return Twitter\Response
```
Fluent interface should be avoided

weierophinney replied:
This... isn't a fluent interface; the upload method is returning a `Response` instance...

Ocramius replied:
My bad, sorry :-\

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     *
+     * @param  Client $httpClient
+     * @return Twitter\Response
+     * @throws Exception If the file can't be opened.
```
More specific exception types needed

weierophinney replied:
They're present, but this docblock hadn't yet been updated to reflect that. Fixing in an upcoming push.

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     */
+    public function __construct($imageFilename = null, $media_type = '')
+    {
+        $this->data['image_filename'] = '';
```
These fields should be properties, not array values

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+        $initResponse = $response->toValue();
+
+        if (! $response->isSuccess()) {
+            throw new Exception\MediaUploadException('Failed to init the upload with Twitter.');
```
More details about the failure are needed. This kind of exception is incredibly hard to debug otherwise.

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+        $params['media_type']  = $this->data['media_type'];
+
+        $response = $this->initUpload($httpClient, $params);
+        $initResponse = $response->toValue();
```
`init`?

weierophinney replied:
What's the question? This is the response to an `INIT` operation.

The refactor I'm preparing to push renames the variable to simply `$response`, as other responses are not captured, however.

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+        $httpClient->setHeaders(['Content-type' => 'application/x-www-form-urlencoded']);
+        $httpClient->setMethod('POST');
+        $httpClient->setParameterPost($payload);
+        $response = $httpClient->send();
```
Assignment is redundant

weierophinney replied:
Refactored this method to raise an exception on error, to be consistent with the other `*Upload()` methods. As such, this line is no longer redundant; the response is tested for success and an exception raised on failure, before returning the more specific `ZendService_Twitter` response.

Ocramius replied:
:+1:

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * @params array $params
+     * @return Twitter\Response
+     */
+    protected function initUpload($httpClient, $params)
```
Type hints needed

Ocramius replied:
`private`

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * @params array $params
+     * @return boolean
+     */
+    protected function appendUpload($httpClient, $params)
```
`private`. type declarations needed

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+            throw new Exception\MediaUploadException('Failed to open the file in the APPEND method.');
+        }
+
+        while (! feof($fileHandle)) {
```
This code is weird - it's not really buffering, so why is it not using `file_get_contents()`? Multiple requests are also being sent...

weierophinney replied:
It _is_ buffering, though. We can only send `$chunkSize` bytes at a time to Twitter (limits they set), so we read that many from the file at a time, sending them in an `APPEND` operation to the API, until the entire file has been read. For large files, being able to do so also saves memory for the process performing the upload, as the entire file is never in memory at one time.

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+        $httpClient->setHeaders(['Content-type' => 'application/x-www-form-urlencoded']);
+        $httpClient->setMethod('POST');
+        $httpClient->setParameterPost($payload);
+        $response = $httpClient->send();
```
Redundant assignment

weierophinney replied:
Fixed.

Ocramius replied:
:+1:

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * @params array $params
+     * @return Twitter\Response
+     */
+    protected function finalizeUpload($httpClient, array $params)
```
types

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+        $returnValue = false;
+        $returnValue = file_exists($fileName);
+
+        if ($returnValue) {
```
`file_exists()` and `is_readable()`?

weierophinney replied:
Looking through the docs, this can just be `is_readable()` (as it combines a `file_exists()` call at the C level). The operations need to be surrounded by an error handler, though, as they can emit an `E_WARNING`.

Updates coming momentarily.

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * @param  string $key
+     * @return mixed
+     */
+    public function __get($key)
```
To be removed

Ocramius commented on library/ZendService/Twitter/Media.php:
```diff
@@ -0,0 +1,251 @@
+     * @param string $key
+     * @param mixed $value
+     */
+    public function __set($key, $value)
```
Absolutely not OK anymore. We used to do this in ZF1, but now this should be an endpoint per property, if mutability is wished. Otherwise, constructor should be preferred

weierophinney commented:
> 5320b6c seems unrelated
> 
> A lot of these commits should be squashed due to their lack of description.

I've now rebased off of develop, and squashed commits.

Ocramius commented:
Thanks @weierophinney! I'll try doing another round of reviews soon.
